crates/ra_cfg/src/lib.rs

```diff
@@ -0,0 +1,43 @@
+//! ra_cfg defines conditional compiling options, `cfg` attibute parser and evaluator
+use ra_syntax::SmolStr;
+use rustc_hash::{FxHashMap, FxHashSet};
+
+mod cfg_expr;
+
+pub use cfg_expr::{parse_cfg, CfgExpr};
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct CfgOptions {
+    atoms: FxHashSet<SmolStr>,
+    features: FxHashSet<SmolStr>,
+    options: FxHashMap<SmolStr, SmolStr>,
+}
```

Comment:
What about `key_values: FxHashSet<(SmolStr, SmolStr)>` ?

Reply:
In cfg, keys are multi-valued. Ie, something like this is a thing:

```rust
$ rustc --cfg key=\"a\" --cfg key=\"b\" main.rs
```

Also, from the experimentation, value-less and valued keys are in different namespaces.

That is, there's a difference between

```rust
$ rustc --cfg key=\"a\" --cfg key main.rs
```

and 

```rust
$ rustc --cfg key=\"a\" main.rs
```

Reply:
Of course, we still need `atoms: FxHashSet<SmolStr>`.
I mean to use `key_values: FxHashSet<(SmolStr, SmolStr)>` instead of `FxHashSet<SmolStr, Vec<SmolStr>>` to speed up lookup key-value cfgs. And as you know, empty `Vec` is invalid here, so I think `Vec` just violates the semantics.

Reply:
>I mean to use key_values: FxHashSet<(SmolStr, SmolStr)> instead of 

Ah, sorry, I've misunderstood that. That's actually really clever, I like this! But it will definitely require a comment in the code explaining what's going on here :-)